### PR TITLE
Bump to C++17

### PR DESCRIPTION
**Pull request by Scrumplex**
This drops optional-bare. See #135

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -29,10 +29,10 @@ set(CMAKE_JAVA_TARGET_OUTPUT_DIR ${PROJECT_BINARY_DIR}/jars)
 ######## Set compiler flags ########
 set(CMAKE_CXX_STANDARD_REQUIRED true)
 set(CMAKE_C_STANDARD_REQUIRED true)
-set(CMAKE_CXX_STANDARD 11)
-set(CMAKE_C_STANDARD 11)
+set(CMAKE_CXX_STANDARD 17)
+set(CMAKE_C_STANDARD 17)
 include(GenerateExportHeader)
-set(CMAKE_CXX_FLAGS "-Wall -pedantic -D_GLIBCXX_USE_CXX11_ABI=0 -fstack-protector-strong --param=ssp-buffer-size=4 ${CMAKE_CXX_FLAGS}")
```
Comment (flowln) [inline CMakeLists.txt:35]: i think you didn't need to change this flag, as it's not about the c++ version we're compiling against, but rather the c++ ABI our libraries will use afaik. shouldn't cause any problems either way, just making a note :)

**Review by flowln — APPROVED** [review]
looks good!

File: CMakeLists.txt (hunk 1 above)
Comment (Scrumplex) [reply CMakeLists.txt:35]: I read that if you really need that flag, GCC would tell you. And it didn't tell me, so I assumed it can go

**Review by txtsd — APPROVED** [review]
Based PR

**Comment by DioEgizio** [comment]
For some reason LGTM doesn't seem to like this pr :thinking:

**Comment by LennyMcLennington** [comment]
> For some reason LGTM doesn't seem to like this pr 🤔

Apparently the compiler it uses doesn't support C++17 or something along those lines. That's what I gathered from the output.

**Comment by txtsd** [comment]
> > For some reason LGTM doesn't seem to like this pr thinking
> 
> Apparently the compiler it uses doesn't support C++17 or something along those lines. That's what I gathered from the output.

That's what it says, but LGTM uses CMake 3.16 so I don't see why it's a problem.

**Review by LennyMcLennington — APPROVED** [review]

**Comment by txtsd** [comment]
lmao @LennyMcLennington you changed it back to 11

**Comment by LennyMcLennington** [comment]
> lmao @LennyMcLennington you changed it back to 11

Read the commit title, description, and diff closely.

I changed it to C11, which is almost identical to C17 (which we have no use for in the first place -- we only wanted to bump to **C++**17). This still does what the title says since I changed the title to reflect the change I made.

**Comment by txtsd** [comment]
Oh yea, my bad. I'm on mobile and misread it.

**Pull request merged**